eth-providers/src/base-provider.ts

```diff
@@ -862,21 +879,77 @@ export abstract class BaseProvider extends AbstractProvider {
     txFeePerGas: (this.api.consts.evm.txFeePerGas as UInt).toBigInt(),
   });
 
-  /**
-   * Estimate gas for a transaction.
-   * @param transaction The transaction to estimate the gas of
-   * @returns The estimated gas used by this transaction
-   */
-  estimateGas = async (transaction: Deferrable<TransactionRequest>): Promise<BigNumber> => {
-    await this.call(transaction);
+  estimateGasV1 = async (transaction: Deferrable<TransactionRequest>): Promise<BigNumber> => {
     const { storageDepositPerByte, txFeePerGas } = this._getGasConsts();
     const gasPrice = (await transaction.gasPrice) || (await this.getGasPrice());
     const storageEntryLimit = BigNumber.from(gasPrice).and(0xffff);
     const storageEntryDeposit = BigNumber.from(storageDepositPerByte).mul(64);
     const storageGasLimit = storageEntryLimit.mul(storageEntryDeposit).div(txFeePerGas);
 
     const resources = await this.estimateResources(transaction);
-    return resources.gas.add(storageGasLimit);
+    return resources.gasLimit.add(storageGasLimit);
+  };
+
+  /**
+   * Estimate gas for a transaction.
+   * @param transaction The transaction to estimate the gas of
+   * @returns The estimated gas used by this transaction
+   */
+  estimateGas = async (transaction: Deferrable<TransactionRequest>): Promise<BigNumber> => {
+    if (!process.env.V2) return this.estimateGasV1(transaction);
+
+    const { usedGas, gasLimit, usedStorage } = await this.estimateResources(transaction);
+
+    const tx = await resolveProperties(transaction);
+    const data = tx.data?.toString() ?? '0x';
+
+    const createParams = [
+      data,
+      toBN(BigNumber.from(tx.value ?? 0)),
+      toBN(gasLimit),
+      toBN(usedStorage.isNegative() ? 0 : usedStorage),
+      (tx.accessList as any) || [],
+    ] as const;
+
+    const callParams = [tx.to!, ...createParams ] as const;
+
+    const extrinsic = tx.to
+      ? this.api.tx.evm.call(...callParams)
+      : this.api.tx.evm.create(...createParams);
+
+    let txFee = await this._estimateGasCost(extrinsic);
+    txFee = txFee.mul(gasLimit).div(usedGas);   // scale it to the same ratio when estimate passing gasLimit
+
+    if (usedStorage.gt(0)) {
+      const storageFee = usedStorage.mul(this._getGasConsts().storageDepositPerByte);
+      txFee = txFee.add(storageFee);
+    }
+
+    const gasPrice = tx.gasPrice && BigNumber.from(tx.gasPrice).gt(0)
+      ? BigNumber.from(tx.gasPrice)
+      : await this.getGasPrice();
+
+    return encodeGasLimit(txFee, gasPrice, gasLimit, usedStorage);
+  };
+
+  _estimateGasCost = async (extrinsic: SubmittableExtrinsic<'promise', ISubmittableResult>) => {
+    const header = await this._getBlockHeader('latest');
+    const isLocalModeFirstBlock = this.localMode && header.parentHash.toHex() === ZERO_BLOCK_HASH;
+    const parentHash = isLocalModeFirstBlock
+      ? header.hash
+      : header.parentHash;
+    const apiAtParentBlock = await this.api.at(parentHash);
```

Comment:
The current hash should be used here. 
The calculation method of the handling fee has changed, not sure if the calculation is correct.
Keep this in mind during testing.

Reply:
I will pay attention to the gas calculation when testing to make sure the it's correct.

Actually as long as it's not too far from the correct gas, it should be fine, since we don't need an exact match here, we usually have `gasLimit = gasEstimation * 120%` to make sure tx pass